File: SJF.py
# --------- SJF Non-Preemptive ---------

class Process:
    def __init__(self, pid, at, bt):
        self.pid = pid
        self.at = at     # Arrival Time
        self.bt = bt     # Burst Time
        self.wt = 0      # Waiting Time
        self.tat = 0     # Turnaround Time
        self.ct = 0      # Completion Time
        self.finished = False

def sjf_scheduling(processes):
    time = 0
    completed = 0
    n = len(processes)

    while completed < n:
        # Ready processes
        ready_queue = [p for p in processes if p.at <= time and not p.finished]

        if ready_queue:
            ready_queue.sort(key=lambda x: (x.bt, x.at))
            current = ready_queue[0]

            time += current.bt
            current.ct = time
            current.tat = current.ct - current.at
            current.wt = current.tat - current.bt
            current.finished = True
            completed += 1
        else:
            time += 1

# Input
n = int(input("Enter number of processes: "))
processes = []

for i in range(n):
    at = int(input(f"Enter Arrival Time for Process {i+1}: "))
    bt = int(input(f"Enter Burst Time for Process {i+1}: "))
    processes.append(Process(i+1, at, bt))

# Run scheduling
sjf_scheduling(processes)

# Output Table
print("\nProcess\tAT\tBT\tWT\tTAT\tCT")
total_wt = total_tat = 0

for p in processes:
    total_wt += p.wt
    total_tat += p.tat
    print(f"P{p.pid}\t{p.at}\t{p.bt}\t{p.wt}\t{p.tat}\t{p.ct}")

print(f"\nAverage Waiting Time = {total_wt / n:.2f}")
print(f"Average Turnaround Time = {total_tat / n:.2f}")

File: SJF.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Process {
    int pid;
    int at;      
    int bt;       
    int rt;       
    int ct = 0;  
    int wt = 0;   
    int tat = 0;  
    bool finished = false;
};

void srtf_scheduling(vector<Process>& processes) {
    int n = processes.size();
    int completed = 0;
    int time = 0;
    int prev = -1; 

    while (completed < n) {
       
        int idx = -1;
        int min_rt = 1e9;
        for (int i = 0; i < n; i++) {
            if (processes[i].at <= time && !processes[i].finished && processes[i].rt < min_rt) {
                min_rt = processes[i].rt;
                idx = i;
            }
        }

        if (idx != -1) {
          
            processes[idx].rt--;
            time++;

            
            if (processes[idx].rt == 0) {
                processes[idx].finished = true;
                completed++;
                processes[idx].ct = time;
                processes[idx].tat = processes[idx].ct - processes[idx].at;
                processes[idx].wt = processes[idx].tat - processes[idx].bt;
            }
        } else {
            
            time++;
        }
    }
}

int main() {
    int n;
    cout << "Enter number of processes: ";
    cin >> n;

    vector<Process> processes(n);
    for (int i = 0; i < n; i++) {
        cout << "Enter Arrival Time for Process " << i + 1 << ": ";
        cin >> processes[i].at;
        cout << "Enter Burst Time for Process " << i + 1 << ": ";
        cin >> processes[i].bt;
        processes[i].pid = i + 1;
        processes[i].rt = processes[i].bt; 
    }

    srtf_scheduling(processes);

    cout << "\nProcess\tAT\tBT\tWT\tTAT\tCT\n";
    double total_wt = 0, total_tat = 0;
    for (const auto& p : processes) {
        total_wt += p.wt;
        total_tat += p.tat;
        cout << "P" << p.pid << "\t" << p.at << "\t" << p.bt
             << "\t" << p.wt << "\t" << p.tat << "\t" << p.ct << "\n";
    }

    cout << "\nAverage Waiting Time = " << total_wt / n << "\n";
    cout << "Average Turnaround Time = " << total_tat / n << "\n";

    return 0;
}

